### DISCUSSION TITLE
Task: let us say goodbye to unsupported Windows releases in NVDA 2024.1

### DISCUSSION THREAD

**Discussion by josephsl** [nvaccess/nvda#15167] [General]
Hi,

The following can be done before or during #12064 work, although I think it shoud be done early in 2024.1 cycle:

### Introduction
Whenever a new Windows version is released (including feature updates for Windows 10 and 11), Microsoft announces end of support dates. For example, Windows 7 was released in 2009, and Microsoft announced january 14, 2020 as end of support date. Similary, Windows 10 was released on July 29, 2015, with end of support for the final feature update (22H2) scheduled for October 14, 2025 for consumers (Windows 10 Enterprise LTSC and Server releases have different end of support dates).

In contrast to earlier releases, Windows 10 and 11 are governed by modern lifecycle policy where feature updates for these Windows releases are supported for a short term. For example, until 21H2 Windows 10 Hoe and Pro feature updates were supported for 18 months, with Windows 11 feature updates supported for at least two years.

As of July 2023, the following Windows versions supported by NVDA are no longer supported:

* Windows 7 (Service Pack 1): end of life, including end of extended security updates (ESU's)
* Windows 8.x: end of life as of January 2023
* Windows 10: feature updates except 1507, 1607, 1809, 21H2, and 22H2 are end of life (feature updates except 22H2 have corresponding Enterprise LTSC version)

### Is your feature request related to a problem? Please describe.
Once #12064 is initiated with a recent Python release (say, 3.11), it becomes necessary to end support for Windows releases listed above as Python itself no longer supports end of life Windows releases. Recent statistics from sources such as Stat Counter and other websites note that majority of systems surveyed are running Windows 10 or 11, with decreasing market share of Windows 7 and 8.x. Also, NV Access noted months ago that devleopers are no longer fixing bugs specific to Windows 7 and 8.x, and coupled with impending Python upgrade, ending support for unsupported Windows releases should be considered.

### Describe the solution you'd like
Review NVDA source code and remove support for unsupported Widows releases as part of NVDA 2024.1 development cycle, preferably before #12064 is initiated. I advise prior to Python upgrade as NVDA users are aware that NV Access is interested in moving to newer Python 3 versions and as newer versions says goodbye to old Windows releases.

#### Proposed supported Windows releases
I propose supporting Windows 10 Version 1607 and Server 2016 or newer. Despite being supported until October 2025, the only reason for running Windows 10 1507 is in mission-critical environments (Windows 10 Enterprise LTSB), and Version 1607 (Enterprise LTSB at this point) and Server 2016 will be supported until October 2026. Or, if we wish to take advantage of UIA improvements out of the box, perhaps requiring Windows 10 1809 and Server 2019 or later would be better (1809 is used in AppVeyor image and Server 2019 will be supported until January 2029).

#### Proposed process
1. Prior to starting NVDA 2024.1 development (provided that the appropriate Python 3 version is chosen), inform NVDA community about impending end of support for unsupported Windows releases.
2. On NV Access server, block NVDA alpha updates on older Windows releases.
3. Review NVDA source code and remove code specific to unsupported Windows releases.
4. Run try builds for several weeks to make sure code changes do not introduce regressions on supported Windows releases.

### Describe alternatives you've considered
None

### Additional context
I think we need to work on persuading add-ons community to end support for unsupported Windows releases, similar to what we did in 2017 when XP and Vista support was ended.

Thanks.

**Comment by Brian1Gaff** [comment]
So then why this very morning did I get a note on my Windows 7 machines 
Microsoft update pending installation and I had to  restart the computer. 
There is no obvious change, though Adobe reader and a couple of tweaks to 
explorer have changed.

There is no other change.
 Brian

-- 
***@***.***
Sent via blueyonder.(Virgin media)
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.
----- Original Message ----- 
From: "Joseph Lee" ***@***.***>
To: "nvaccess/nvda" ***@***.***>
Cc: "Subscribed" ***@***.***>
Sent: Wednesday, July 12, 2023 3:45 PM
Subject: [nvaccess/nvda] Task: let us say goodbye to unsupported Windows 
releases in NVDA 2024.1 (Issue #15135)

**Comment by enessaribas** [comment]
Windows 7 hasn't been supported since 2020 for consumers.  Didn't NVAccess announce on the nvdadevel list that support would be dropped for 7/8x from 2024.1?

**Comment by josephsl** [comment]
Hi, yes. But even then I think a reminder just before starting 2024.1 dev cycle would be good in case people did not see earlier announcements. Thanks.

**Comment by enessaribas** [comment]
would dropping support for all Windows releases up to 1809 enable performance enhancements or features not currently existing in NVDA?

**Comment by josephsl** [comment]
Hi,

Windows 10 1809 (October 2018 Update) gives us:

* Native IUIAutomation6 interface
* Cleaner app modules, especially File Explorer and emoji panel
* Running on Windows 10 on ARM without checking if ARM64 remote helper must be initialized depending on Windows release in use (Windows 10 on ARM was introduced in Version 1709)

In short, it should result in less clutter in NVDA source code, particularly for app modules.

Performance depends on Python 3 - more recent releases such as 3.10 and 3.11 offer speed improvements which may or may not affect NVDA (parts of NVDA are written in C++ anyway).

Thanks.

**Comment by seanbudd** [comment]
I don't think it is a pressing issue to remove redundant code in 2024.1 dev cycle.
As usual with NVDA code, we clean up as we go, rather than spend time refactoring the entire codebase at one time.

**Comment by seanbudd** [comment]
We also do not think we should be removing code related to Windows 10 - we are only dropping support for Win 7/8

**Comment by josephsl** [comment]
Hi,

October 2023 update: #12064 is under development (a branch different from master), with Windows 8.1 being the earliest supported release. My suggestion to not offer alpha builds to folks running Windows 7 SP1 and 8.0 (and server counterparts) before Python 3.11 work is merged to master stands. As for ending support for Windows 8.1 and Server 2012 R2, that will depend on Python support lifecycle.

Based on Python 3.11 development progress (October 2023), the following code can be removed (or should be removed in the future) if supporting Windows 8.1 and later:

* Detecting older style system tray and Start menu found in Windows 7 SP1
* Checking for IUIAutomation (Windows 7) interface - we can start with IUIAutomation3
* Removing Windows 7 and 8.0 constants from winVersion module
* In touch handler, checking for presence of touchscreen hardware rather than looking for both hardware and the OS release
* Updates to winUser and winKernel wrappers
* Accompanying user guide edits
* Anything else we need to look at

Thanks.

**Reply by LeonarddeR** [reply]
I created https://github.com/nvaccess/nvda/issues/15577 to drop support for the EOA Proxy

**Comment by beqabeqa473** [comment]
Just for interest, is there any reason to keep 8.1 compatibility if it
is already out of support?

On 10/4/23, Joseph Lee ***@***.***> wrote:
> Hi,
>
> October 2023 update: #12064 is under development (a branch different from
> master), with Windows 8.1 being the earliest supported release. My
> suggestion to not offer alpha builds to folks running Windows 7 SP1 and 8.0
> (and server counterparts) before Python 3.11 work is merged to master
> stands. As for ending support for Windows 8.1 and Server 2012 R2, that will
> depend on Python support lifecycle.
>
> Based on Python 3.11 development progress (October 2023), the following code
> can be removed (or should be removed in the future) if supporting Windows
> 8.1 and later:
>
> * Detecting older style system tray and Start menu found in Windows 7 SP1
> * Checking for IUIAutomation (Windows 7) interface - we can start with
> IUIAutomation3
> * Removing Windows 7 and 8.0 constants from winVersion module
> * In touch handler, checking for presence of touchscreen hardware rather
> than looking for both hardware and the OS release
> * Updates to winUser and winKernel wrappers
> * Accompanying user guide edits
> * Anything else we need to look at
>
> Thanks.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/discussions/15167#discussioncomment-7190266
> You are receiving this because you are subscribed to this thread.
>
> Message ID:
> ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by josephsl** [comment]
Hi, Python 3.11 (and 3.12) docs say Python supports Windows 8.1 and later. Thanks.

**Comment by lukaszgo1** [comment]
>     * Detecting older style system tray

What code you mean exactly?

> 
>     * Removing Windows 7 and 8.0 constants from winVersion module

We use these constants in `winVersion.isSupportedOS`, so they should remain.

**Comment by josephsl** [comment]
Hi,
System tray: I think there is code in File Explorer app module to deal with some Windows 7 code, and I thought it wsa about system tray (perhaps not).

winVersion constants: winVersion.isSupportedOS() returns winVersion.WIN81 or later (with Python 3.11 work), so Windows 7 and 8.0 constants may not matter unless we need to keep them for historical reasons.

Thanks.

**Comment by Brian1Gaff** [comment]
Are there stats that NV Access has containing the numbers using Windows 7 and the Windows 8?
 This is coming from a Windows 7 machine.
 I do have special reasons for using it, but am perfectly prepared to run older NVDA if Support is withdrawn, assuming a miniscule number of users exists. At the moment because Narrator really only became viable in later Windows10, there is a case for keeping a version for Windows 7 and 8, as this is the only game in town.
 
However if you are sure you wish to stop supporting older windows, there are probably app modules that can be removed, not just detection of certain forms of objects. One is what I am using right now, Outlook Express, as I do not think the  kludged version of this product is still being maintained since the demise of Internet Explorer, whose code is used by the software.
 Brian

-- 
***@***.*** 
Sent via blueyonder.(Virgin media) 
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.

**Reply by seanbudd** [reply]
Hi Brian - stats on Win 7 and 8 usage will not affect the decision to drop support.

Right now there is about 6% of users on Win 7, and less than 1% on Win 8
http://nvda-project.org/userStats?date=&numDays=7&stat=simple_os_version

**Comment by lukaszgo1** [comment]
> Are there stats that NV Access has containing the numbers using Windows 7 and the Windows 8? This is coming from a Windows 7 machine. I do have special reasons for using it, but am perfectly prepared to run older NVDA if Support is withdrawn, assuming a miniscule number of users exists. At the moment because Narrator really only became viable in later Windows10, there is a case for keeping a version for Windows 7 and 8, as this is the only game in town.

I assume, though NV Access didn't mention this explicitly, that the situation would be similar to what has happened for NVDA 2017.3 i.e. a link to download 2023.3 on the downloads page with a disclaimer that this version is not in active support.


> However if you are sure you wish to stop supporting older windows, there are probably app modules that can be removed, not just detection of certain forms of objects. One is what I am using right now, Outlook Express, as I do not think the kludged version of this product is still being maintained since the demise of Internet Explorer, whose code is used by the software.

We had this discussion in #12218, and I don't think much has  changed since then. For reference Outlook Express can still be installed on Windows 10, and at least in the Polish blind community there are people using it for whatever reason.

**Comment by zstanecic** [comment]
hi Łukasz,

Some polish people in the polish community use outlook express because they don’t want to learn new new ways how to use something.

That is also a thing with the old speech synthesizers.

People can stick with the old things, but for how long?

**Comment by josephsl** [comment]
Hi,

Post-Python 3.11 update: now that #12064 is here, I propose creating issue/PR pairs to work on removing Windows 7 and 8.0 support in places not touched by that issue.

In addition, the following may need to be looked at:

* App module handler: simplify store app detection by checking for immersive app function value.
* Vision enhancement provider/screen curtain: see if we can use API's from Windows 8.1 and later.
* Other things.

I know some of these will also touch the user guide, so please take care of it as well.

Thanks.
